datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -1991,6 +1992,71 @@ pub fn array_intersect(args: &[ArrayRef]) -> Result<ArrayRef> {
     }
 }
 
+pub fn general_array_distinct<OffsetSize: OffsetSizeTrait>(
+    array: &GenericListArray<OffsetSize>,
+    field: &FieldRef,
+) -> Result<ArrayRef> {
+    let dt = array.value_type();
+    let mut offsets = vec![OffsetSize::usize_as(0)];
+    let mut new_arrays = vec![];
+    let converter = RowConverter::new(vec![SortField::new(dt.clone())])?;
+    // distinct for each list in ListArray
+    for arr in array.iter().flatten() {
+        let values = converter.convert_columns(&[arr])?;
```

Comment:
why not distinct array in columnar way, `arr` has only one column, using row format need extra encoding and decoding

Reply:
> why not distinct array in columnar way, `arr` has only one column, using row format need extra encoding and decoding

It is a great to distinct array without row converter, but I don't think we can do that without downcast to exact arr then do the distinction. Is there any way to avoid row converters?

Reply:
I also don't have another idea other than downcast arr, I was just wondering if it is worth to downcast to exact arr.